obstore/python/obstore/auth/earthdata.py

```diff
@@ -6,157 +6,419 @@
-    )
-    store = S3Store("bucket_name", credential_provider=credential_provider)
-    ```
+    Examples:
+        ```py
+        import obstore
+        from obstore.auth.earthdata import NasaEarthdataCredentialProvider
+
+        # Obtain an S3 credentials URL and an S3 data/download URL, typically
+        # via metadata returned from a NASA CMR collection or granule query.
+        credentials_url = "https://data.ornldaac.earthdata.nasa.gov/s3credentials"
+        data_url = "s3://ornl-cumulus-prod-protected/gedi/GEDI_L4A_AGB_Density_V2_1/data/GEDI04_A_2024332225741_O33764_03_T01289_02_004_01_V002.h5"
+        data_prefix_url, filename = data_url.rsplit("/", 1)
+
+        # Since no Earthdata credentials are specified, environment variables or netrc
+        # will be used to locate them in order to obtain S3 credentials from the URL.
+        credential_provider = NasaEarthdataAsyncCredentialProvider(credentials_url)
+        store = obstore.store.from_url(data_prefix_url, credential_provider=credential_provider)
+
+        # Download the file by streaming chunks
+        try:
+            result = await obstore.get_async(store, filename)
+            with open(filename, "wb") as f:
+                async for chunk in aiter(result):
+                    f.write(chunk)
+        finally:
+            await credential_provider.close()
+        ```
 
     [NASA Earthdata]: https://www.earthdata.nasa.gov/
+    [in-region direct S3 access]: https://archive.podaac.earthdata.nasa.gov/s3credentialsREADME
+
     """  # noqa: E501
 
     config: S3Config
 
     def __init__(
         self,
-        username: str,
-        password: str,
+        credentials_url: str,
+        *,
+        host: str | None = None,
+        auth: str | tuple[str, str] | None = None,
+        session: aiohttp.ClientSession | aiohttp_retry.RetryClient | None = None,
     ) -> None:
-        """Create a new NasaEarthdataAsyncCredentialProvider.
+        """Construct a new NasaEarthdataAsyncCredentialProvider.
 
-        Args:
-            username: Username to NASA Earthdata.
-            password: Password to NASA Earthdata.
+        This credential provider uses `aiohttp`, and will error if that cannot be
+        imported.
 
-        """
-        from aiohttp import BasicAuth, ClientSession
+        Refer to
+        [NasaEarthdataCredentialProvider][obstore.auth.earthdata.NasaEarthdataCredentialProvider.__init__]
+        for argument explanations.
 
-        # Pass region default
+        """
         self.config = {"region": "us-west-2"}
-        self.session = ClientSession(auth=BasicAuth(username, password))
+        self._session: aiohttp.ClientSession | aiohttp_retry.RetryClient | None = (
+            session or default_aiohttp_session()
+        )
+        # Avoid closing a user-supplied session (the user is responsible for that)
+        self._close = None if session else self._session.close
+        self._get_credentials_async = _make_get_credentials_async(
```

Comment:
Instead of a factory function that returns another function; can we just move the body of `def _make_get_credentials_async` into a class instance method? Seems like that would be a lot simpler and we wouldn't need to deal with closures.